File: ts-nodejs/test/skills.test.ts
import { describe, expect, test } from "./bun-shim.ts";
import { mkdirSync, rmSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { collectSkills, parseFrontmatter, parseFrontmatterFromBytes, scanSkills, sortSkills } from "../src/core/skills.ts";

const bytes = (text: string): Uint8Array => new TextEncoder().encode(text);
const utf8 = (text: string): Uint8Array => new TextEncoder().encode(text);

const fixture = (): { dir: string; skill: (id: string, body: Uint8Array | string) => void } => {
  const dir = join(process.env["TMP"] ?? ".", `kpt-skills-${Math.random().toString(36).slice(2)}`);
  mkdirSync(dir, { recursive: true });
  return {
    dir,
    skill(id, body) {
      const sub = join(dir, id);
      mkdirSync(sub, { recursive: true });
      writeFileSync(join(sub, "SKILL.md"), typeof body === "string" ? utf8(body) : body);
    },
  };
};

const fm = (text: string): [string | null, string | null] => parseFrontmatterFromBytes(bytes(text));

describe("frontmatter parser (SPEC 21.2)", () => {
  test("frontmatter_name_description_and_quotes, ported from rust/src/skills.rs", () => {
    const fx = fixture();
    fx.skill("a", '---\nname: Alpha\ndescription: "Does things"\n---\nbody');
    fx.skill("b", "---\ndescription: 'Only desc'\n---\n");
    fx.skill("c", "no frontmatter at all");
    mkdirSync(join(fx.dir, "empty-no-skillmd"), { recursive: true });

    const out = collectSkills(fx.dir, "Test");
    expect(out).toHaveLength(3);

    const a = out.find((s) => s.id === "a");
    expect(a?.name).toBe("Alpha");
    expect(a?.description).toBe("Does things");

    const b = out.find((s) => s.id === "b");
    expect(b?.name).toBe("b"); // falls back to the directory name
    expect(b?.description).toBe("Only desc");

    const c = out.find((s) => s.id === "c");
    expect(c?.name).toBe("c");
    expect(c?.description).toBe("");
    rmSync(fx.dir, { recursive: true, force: true });
  });

  test("bom_and_non_utf8_bytes_are_tolerated, ported from rust/src/skills.rs", () => {
    const fx = fixture();
    fx.skill("bom", utf8("﻿---\nname: Bom\ndescription: x\n---\n"));
    const gbk = [...utf8("---\nname: Gbk\ndescription: "), 0xd6, 0xd0, 0xce, 0xc4, ...utf8("\n---\n")];
    fx.skill("gbk", Uint8Array.from(gbk));

    const out = collectSkills(fx.dir, "Test");
    expect(out).toHaveLength(2);
    expect(out.find((s) => s.id === "bom")?.name).toBe("Bom");
    const gbkSkill = out.find((s) => s.id === "gbk");
    expect(gbkSkill?.name).toBe("Gbk"); // lossy decode keeps the ASCII fields
    expect(gbkSkill?.description.includes("\uFFFD")).toBe(true); // GBK bytes decode lossily to U+FFFD
    rmSync(fx.dir, { recursive: true, force: true });
  });

  test("the fence must be the first line and may carry a BOM or padding", () => {
    expect(fm("﻿---\nname: A\n---\n")[0]).toBe("A");
    expect(fm("  ---  \nname: A\n---\n")[0]).toBe("A");
    expect(fm("﻿﻿---\nname: A\n---\n")[0]).toBe("A");
    expect(fm("----\nname: A\n---\n")[0]).toBeNull();
    expect(fm("---x\nname: A\n---\n")[0]).toBeNull();
  });

  test("keys stop at the closing fence, which may be indented", () => {
    expect(fm("---\nname: A\n  ---\nname: B\n")[0]).toBe("A");
    expect(fm("---\nname: A\n")[0]).toBe("A"); // no closing fence at all
  });

  test("the first value wins and only the first colon splits", () => {
    expect(fm("---\nname: One\nname: Two\n")[0]).toBe("One");
    expect(fm("---\ndescription: see http://x/y\n")[1]).toBe("see http://x/y");
  });

  test("quote stripping is trim_matches, not a matching pair", () => {
    expect(fm('---\ndescription: "quoted"\n')[1]).toBe("quoted");
    expect(fm("---\ndescription: 'single'\n")[1]).toBe("single");
    expect(fm('---\ndescription: ""mixed""\n')[1]).toBe("mixed");
    expect(fm('---\ndescription: "a" and "b"\n')[1]).toBe('a" and "b');
    expect(fm("---\ndescription:\n")[1]).toBe("");
  });

  test("an indented nested key can be picked up, as the Rust comment warns", () => {
    expect(fm("---\nmetadata:\n  name: nested\n")[0]).toBe("nested");
    expect(fm("---\ndescription: >-\n  folded\n")[1]).toBe(">-");
  });

  test("the parser stops at 4 KiB of the bytes it is handed", () => {
    const long = "---\ndescription: " + "x".repeat(5000) + "\nname: Late\n---\n";
    const parsed = parseFrontmatterFromBytes(bytes(long));
    expect(parsed[1]).toHaveLength(4096 - "---\ndescription: ".length);
    expect(parsed[0]).toBeNull();
  });

  test("the 4 KiB window also holds through the real file path (SPEC 21.2)", () => {
    const fx = fixture();
    // This pins the file path end to end (readFileBytes + parser agree on the
    // window). It cannot pin the *physical* read bound: the parser caps at 4096
    // itself, so a whole-file read is indistinguishable through any public API —
    // that half of Rust's `take(4096)` is a memory-safety property, code-reviewed
    // only.
    fx.skill("big", "---\ndescription: " + "x".repeat(5000) + "\nname: Late\n---\n");
    const [name, description] = parseFrontmatter(join(fx.dir, "big", "SKILL.md"));
    expect(description).toHaveLength(4096 - "---\ndescription: ".length);
    expect(name).toBeNull();
    rmSync(fx.dir, { recursive: true, force: true });
  });
});

describe("ordering (SPEC 21.2)", () => {
  test("source compares by code point, name case-insensitively", () => {
    const sorted = sortSkills([
      { id: "z", name: "zeta", description: "", source: "Kimi Code" },
      { id: "a", name: "Alpha", description: "", source: "Kimi Code" },
      { id: "b", name: "beta", description: "", source: "Agents" },
      { id: "p", name: "x", description: "", source: "Plugin: zzz" },
    ]);
    expect(sorted.map((s) => `${s.source}/${s.name}`)).toEqual([
      "Agents/beta",
      "Kimi Code/Alpha",
      "Kimi Code/zeta",
      "Plugin: zzz/x",
    ]);
  });

  test("case-insensitive means code points, not a locale collation", () => {
    const sorted = sortSkills([
      { id: "1", name: "b", description: "", source: "S" },
      { id: "2", name: "_", description: "", source: "S" },
      { id: "3", name: "A", description: "", source: "S" },
    ]);
    // lowercased code points: _ (5F) < a (61) < b (62)
    expect(sorted.map((s) => s.name)).toEqual(["_", "A", "b"]);
  });

  test("scanning a real machine never throws", () => {
    expect(Array.isArray(scanSkills(process.env))).toBe(true);
  });
});

File: go/internal/core/skills_test.go
package core

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func skillFixture(t *testing.T) string {
	t.Helper()
	return t.TempDir()
}

func writeSkill(t *testing.T, dir, id, body string) {
	t.Helper()
	sub := filepath.Join(dir, id)
	if err := os.MkdirAll(sub, 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(sub, "SKILL.md"), []byte(body), 0o644); err != nil {
		t.Fatal(err)
	}
}

func writeSkillBytes(t *testing.T, dir, id string, body []byte) {
	t.Helper()
	sub := filepath.Join(dir, id)
	if err := os.MkdirAll(sub, 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(sub, "SKILL.md"), body, 0o644); err != nil {
		t.Fatal(err)
	}
}

func fm(t *testing.T, text string) (*string, *string) {
	t.Helper()
	return ParseFrontmatterFromBytes([]byte(text))
}

func TestCollectSkillsFrontmatterAndQuotes(t *testing.T) {
	dir := skillFixture(t)
	writeSkill(t, dir, "a", "---\nname: Alpha\ndescription: \"Does things\"\n---\nbody")
	writeSkill(t, dir, "b", "---\ndescription: 'Only desc'\n---\n")
	writeSkill(t, dir, "c", "no frontmatter at all")
	if err := os.MkdirAll(filepath.Join(dir, "empty-no-skillmd"), 0o755); err != nil {
		t.Fatal(err)
	}

	out := CollectSkills(dir, "Test")
	if len(out) != 3 {
		t.Fatalf("collect = %d skills, want 3", len(out))
	}
	byID := map[string]SkillInfo{}
	for _, s := range out {
		byID[s.ID] = s
	}
	if byID["a"].Name != "Alpha" || byID["a"].Description != "Does things" {
		t.Errorf("skill a = %+v", byID["a"])
	}
	if byID["b"].Name != "b" || byID["b"].Description != "Only desc" {
		t.Errorf("skill b (name falls back to dir) = %+v", byID["b"])
	}
	if byID["c"].Name != "c" || byID["c"].Description != "" {
		t.Errorf("skill c = %+v", byID["c"])
	}
}

func TestBOMAndNonUTF8BytesAreTolerated(t *testing.T) {
	dir := skillFixture(t)
	writeSkill(t, dir, "bom", "\uFEFF---\nname: Bom\ndescription: x\n---\n")
	var gbk []byte
	gbk = append(gbk, []byte("---\nname: Gbk\ndescription: ")...)
	gbk = append(gbk, 0xD6, 0xD0, 0xCE, 0xC4) // "中文" in GBK
	gbk = append(gbk, []byte("\n---\n")...)
	writeSkillBytes(t, dir, "gbk", gbk)

	out := CollectSkills(dir, "Test")
	if len(out) != 2 {
		t.Fatalf("collect = %d skills, want 2", len(out))
	}
	byID := map[string]SkillInfo{}
	for _, s := range out {
		byID[s.ID] = s
	}
	if byID["bom"].Name != "Bom" {
		t.Errorf("bom skill name = %q", byID["bom"].Name)
	}
	if byID["gbk"].Name != "Gbk" {
		t.Errorf("lossy decode keeps the ASCII fields, got %q", byID["gbk"].Name)
	}
	if !strings.Contains(byID["gbk"].Description, "\uFFFD") {
		t.Errorf("GBK bytes must become U+FFFD (written as an escape per REVIEW-M1 Major D), got %q",
			byID["gbk"].Description)
	}
}

func TestFrontmatterFenceRules(t *testing.T) {
	if n, _ := fm(t, "\uFEFF---\nname: A\n---\n"); n == nil || *n != "A" {
		t.Error("a BOM before the fence is tolerated")
	}
	if n, _ := fm(t, "  ---  \nname: A\n---\n"); n == nil || *n != "A" {
		t.Error("padding around the fence is tolerated")
	}
	if n, _ := fm(t, "\uFEFF\uFEFF---\nname: A\n---\n"); n == nil || *n != "A" {
		t.Error("multiple BOMs are stripped")
	}
	if n, _ := fm(t, "----\nname: A\n---\n"); n != nil {
		t.Error("---- is not a fence")
	}
	if n, _ := fm(t, "---x\nname: A\n---\n"); n != nil {
		t.Error("---x is not a fence")
	}
}

func TestFrontmatterClosingFence(t *testing.T) {
	if n, _ := fm(t, "---\nname: A\n  ---\nname: B\n"); n == nil || *n != "A" {
		t.Error("an indented closing fence ends the block; first value wins")
	}
	if n, _ := fm(t, "---\nname: A\n"); n == nil || *n != "A" {
		t.Error("no closing fence at all still parses")
	}
}

func TestFrontmatterFirstValueWinsAndFirstColonSplits(t *testing.T) {
	if n, _ := fm(t, "---\nname: One\nname: Two\n"); n == nil || *n != "One" {
		t.Error("the first name wins")
	}
	if _, d := fm(t, "---\ndescription: see http://x/y\n"); d == nil || *d != "see http://x/y" {
		t.Error("only the first colon splits")
	}
}

func TestFrontmatterQuoteStrippingIsTrimMatches(t *testing.T) {
	if _, d := fm(t, "---\ndescription: \"quoted\"\n"); d == nil || *d != "quoted" {
		t.Error("double quotes stripped")
	}
	if _, d := fm(t, "---\ndescription: 'single'\n"); d == nil || *d != "single" {
		t.Error("single quotes stripped")
	}
	if _, d := fm(t, "---\ndescription: \"\"mixed\"\"\n"); d == nil || *d != "mixed" {
		t.Error("repeated quotes stripped")
	}
	if _, d := fm(t, "---\ndescription: \"a\" and \"b\"\n"); d == nil || *d != "a\" and \"b" {
		t.Error("inner quotes stay")
	}
	if _, d := fm(t, "---\ndescription:\n"); d == nil || *d != "" {
		t.Error("empty value reads as empty string")
	}
}

func TestFrontmatterKnownLimits(t *testing.T) {
	// An indented nested key can be picked up when the top-level one is
	// absent, and folded scalars show the raw indicator — the Rust comment
	// documents both.
	if n, _ := fm(t, "---\nmetadata:\n  name: nested\n"); n == nil || *n != "nested" {
		t.Error("indented nested name is picked up (documented limit)")
	}
	if _, d := fm(t, "---\ndescription: >-\n  folded\n"); d == nil || *d != ">-" {
		t.Error("folded scalar shows the raw indicator (documented limit)")
	}
}

func TestFrontmatterStopsAt4KiB(t *testing.T) {
	long := "---\ndescription: " + strings.Repeat("x", 5000) + "\nname: Late\n---\n"
	name, description := ParseFrontmatterFromBytes([]byte(long))
	if name != nil {
		t.Error("the name beyond the 4 KiB window must not be seen")
	}
	if description == nil || len(*description) != 4096-len("---\ndescription: ") {
		t.Errorf("description is clipped to the window, got %d bytes", len(*description))
	}
}

func TestFrontmatter4KiBThroughFilePath(t *testing.T) {
	dir := skillFixture(t)
	writeSkill(t, dir, "big", "---\ndescription: "+strings.Repeat("x", 5000)+"\nname: Late\n---\n")
	name, description := ParseFrontmatter(filepath.Join(dir, "big", "SKILL.md"))
	if name != nil {
		t.Error("the name beyond the 4 KiB window must not be seen")
	}
	if description == nil || len(*description) != 4096-len("---\ndescription: ") {
		t.Errorf("file path reads agree with the parser on the window, got %d", len(*description))
	}
}

func TestSortSkills(t *testing.T) {
	sorted := SortSkills([]SkillInfo{
		{ID: "z", Name: "zeta", Source: "Kimi Code"},
		{ID: "a", Name: "Alpha", Source: "Kimi Code"},
		{ID: "b", Name: "beta", Source: "Agents"},
		{ID: "p", Name: "x", Source: "Plugin: zzz"},
	})
	want := []string{"Agents/beta", "Kimi Code/Alpha", "Kimi Code/zeta", "Plugin: zzz/x"}
	for i, w := range want {
		if got := sorted[i].Source + "/" + sorted[i].Name; got != w {
			t.Errorf("sorted[%d] = %q, want %q", i, got, w)
		}
	}

	sorted2 := SortSkills([]SkillInfo{
		{ID: "1", Name: "b", Source: "S"},
		{ID: "2", Name: "_", Source: "S"},
		{ID: "3", Name: "A", Source: "S"},
	})
	want2 := []string{"_", "A", "b"} // lowercased code points: 5F < 61 < 62
	for i, w := range want2 {
		if sorted2[i].Name != w {
			t.Errorf("case-insensitive order [%d] = %q, want %q", i, sorted2[i].Name, w)
		}
	}
}

func TestScanSkillsRealMachine(t *testing.T) {
	// A scan over the real machine must never throw; the result may be empty.
	_ = ScanSkills(nil)
}
